plugins/buildingplan-planner.cpp

```diff
@@ -1,552 +1,922 @@
+        case df::enums::building_type::GrateFloor:
+        case df::enums::building_type::BarsVertical:
+        case df::enums::building_type::BarsFloor:
+        case df::enums::building_type::Cabinet:
+        case df::enums::building_type::Box:
+        case df::enums::building_type::Weaponrack:
+        case df::enums::building_type::Statue:
+        case df::enums::building_type::Slab:
+        case df::enums::building_type::Table:
+        case df::enums::building_type::WindowGlass:
+        case df::enums::building_type::AnimalTrap:
+        case df::enums::building_type::Chain:
+        case df::enums::building_type::Cage:
+        case df::enums::building_type::TractionBench:
+            return true;
+        default:
+            return false;
+    }
 }
 
-void Planner::adjustMaxQuality(df::building_type type, int amount)
+Planner::ItemFiltersWrapper Planner::getItemFilters(BuildingTypeKey key)
 {
-    auto max_quality = &getDefaultItemFilterForType(type)->max_quality;
-    *max_quality = static_cast<df::item_quality>(*max_quality + amount);
+    static std::vector<ItemFilter> empty_vector;
+    static const ItemFiltersWrapper empty_ret(empty_vector);
 
-    boundsCheckItemQuality(max_quality);
-    auto min_quality = &getDefaultItemFilterForType(type)->min_quality;
-    if (*max_quality < *min_quality)
-        (*min_quality) = *max_quality;
+    size_t nfilters = getNumFilters(key);
+    if (nfilters < 1)
+        return empty_ret;
+    while (default_item_filters[key].size() < nfilters)
+        default_item_filters[key].push_back(ItemFilter());
+    return ItemFiltersWrapper(default_item_filters[key]);
 }
 
-void Planner::enableQuickfortMode()
+// precompute a bitmask with bad item flags
+struct BadFlags
 {
-    saved_planmodes = planmode_enabled;
-    for_each_(planmode_enabled, enable_quickfort_fn);
+    uint32_t whole;
 
-    quickfort_mode = true;
-}
+    BadFlags()
+    {
+        df::item_flags flags;
+        #define F(x) flags.bits.x = true;
+        F(dump); F(forbid); F(garbage_collect);
+        F(hostile); F(on_fire); F(rotten); F(trader);
+        F(in_building); F(construction); F(in_job);
+        F(owned); F(in_chest); F(removed); F(encased);
+        #undef F
+        whole = flags.whole;
+    }
+};
 
-void Planner::disableQuickfortMode()
+static bool itemPassesScreen(df::item * item)
 {
-    planmode_enabled = saved_planmodes;
-    quickfort_mode = false;
+    static BadFlags bad_flags;
+    return !(item->flags.whole & bad_flags.whole)
+        && !item->isAssignedToStockpile()
+        // TODO: make this configurable
+        && !(item->getType() == df::item_type::BOX && item->isBag());
 }
 
-bool Planner::inQuickFortMode() { return quickfort_mode; }
+static bool matchesFilters(df::item * item,
+                           df::job_item * job_item,
+                           const ItemFilter & item_filter)
+{
+    if (job_item->item_type > -1 && job_item->item_type != item->getType())
```

Comment:
these checks feel like they should be part of isSuitableItem(), but the function signature would need to be changed. Is there any reason basic attributes like item type are *not* checked in isSuitableItem()? or does isSuitableItem() assume that the item vector will always be set correctly, and that all items in that vector will have the same type?